scripts/abins/abinsalgorithm.py

```diff
@@ -1065,3 +1067,45 @@ def _check_threads(message_end=""):
             return  # Pool() will set a sensible default
         if not (1 <= threads <= multiprocessing.cpu_count()):
             raise RuntimeError("Invalid number of threads for parallelisation over atoms" + message_end)
+
+
+def validate_e_init(*, e_init: str, energy_units: str, instrument_name: str) -> dict[str, str]:
+    """Validator for direct-geometry incident energy parameters
+
+    Check that algorithm inputs specify a valid energy for the given instrument
+    and return appropriate issues dict for .validateInputs() method.
+
+    Note that parameters are strings; this is the form they are received from
+    algorithm parameters.
+
+    Args:
+        e_init: IncidentEnergy for Abins2D
+        energy_units: 'meV' or 'cm-1'
+        instrument_name:
+            A valid Abins2D instrument. This must be validated elsewhere,
+            with a corresponding entry in abins.parameters.instruments.
+    """
+
+    issues = {}
+
+    if not isinstance(float(e_init), numbers.Real):
+        issues["IncidentEnergy"] = "Incident energy must be a real number"
+        return issues
+
+    if "max_wavenumber" in abins.parameters.instruments[instrument_name]:
+        max_energy = abins.parameters.instruments[instrument_name]["max_wavenumber"]
+    else:
+        max_energy = abins.parameters.sampling["max_wavenumber"]
+
+    match energy_units:
+        case "meV":
+            max_energy = max_energy / MILLI_EV_TO_WAVENUMBER
+        case "cm-1":
+            max_energy = max_energy
+        case _:
+            raise ValueError(f"Invalid energy unit: {energy_units}")
```

Comment:
```suggestion
            issues["EnergyUnits"] = f"Invalid energy unit: {energy_units}"
            return issues
```
I think this would be more in keeping with how the validation normally works.
You should also add a test case with an invalid energy unit name.

I know it's 'impossible' to get to this state since `EnergyUnits` has a string list validator on it, but you might change something in one place and not the other one day.

Reply:
🤦 Thanks for catching, this is what I originally intended! I'll sort out the test as well.

Agreed that where it is convenient to catch _and test_ these "impossible" cases it doesn't hurt to do so. (Another approach would be to use a runtime `assert`.) Things get tricky (and coverage trackers get grumpy) when the code is truly unreachable!